DANS-KNAW/easy-deposit-api
# EASY-2170 : locks for submits and POST/PUT uploads; EASY-2081 : internal service error

Fixes 
EASY-2170 : locks for submits and POST/PUT uploads;
EASY-2081 : internal service error

## message for squash merge
* combines both staged directories into a single one
* implements optimistic locking also for 
  * [x] PUT uploads
  * [x] submits
* fetching `StateInfo` of a submitted deposit won't cause an internal error when properties in (or pointing to) the ingest-flow-inbox are not found (not yet or no longer). The state available in the draft area will be returned and the missing property logged.

## Where should the reviewer @DANS-KNAW/easy start?

* [x] see comments on #162

#### How should this be manually tested?

Requirements
* this pull request deployed on deasy
* a small file and a large file (a few GB) to have some time to start a concurrent upload/submit
* the UUID of a draft dataset that is ready to receive uploads
* two browser windows with the deposit-ui ready for a POST upload (or submit with a large file in the bag)
* a curl command prepared for a PUT upload:
  ```
  curl -v -u user001:user001 -H 'Content-Type: text/plain' -H 'Content-Disposition': 'form-data; filename=File3.txt' -F data=@/PATH-TO/XXX.txt -X PUT "http://deasy.dans.knaw.nl/deposit-api/deposit/<UUID>/file/filepath"
  ```
  for an invalid cookie: replace the -u option with `-c 'scentry.auth.default.user=ra.bar.bera'`
* a `tail -f` for deposit api
* monitor files and their sizes:
  ```
  watch ls -la /var/opt/dans.knaw.nl/tmp/easy-deposit-api/*
  ```

Observations:
* a POST upload creates a file in `multipart-location` until the status bar reaches 100%, then a directory is created in `staged` and a log-line appears that the request is received
  * another POST upload fails with a warning to try again later
  * same for another PUT upload
* a PUT upload
  * with valid credentials and the UUID of a draft deposit immediately creates a directory in `staged`
  * with an invalid cookie the server immediately returns the `missing, invalid or expired credentials` but curl apparently keeps writing the complete file

#### Related pull requests on github
https://github.com/DANS-KNAW/easy-dtap/pull/344
